### PR TITLE
Challenge_sum_even_odd

### LINKED ISSUE #41: Challenge_sum_even_odd

### PR DESCRIPTION
**Pull request by ermishina**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/sum_even_and_odd.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+"""This module provides a function sum_even_and_odd
+that calculates the sums of positive and negative,
+even and odd numbers in a given list.
+"""
+
+
+def sum_even_and_odd(numbers):
```
Comment (Azizsin7) [inline solutions/sum_even_and_odd.py]: add annotations

File: solutions/sum_even_and_odd.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+"""This module provides a function sum_even_and_odd
```
Comment (Azizsin7) [inline solutions/sum_even_and_odd.py:4]: add 
shebang line: #!/usr/bin/env python3
encoding declaration: # -*- coding: utf-8 -*-

File: solutions/sum_even_and_odd.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    Raises:
+        ValueError: If input is not a list or contains non-numeric values.
+
+    Examples:
+    >>> sum_even_and_odd([1, -2, 3, 4, -5, 0])
+    {'positive_even': 4, 'positive_odd': 4, 'negative_even': -2, 'negative_odd': -5}
+    >>> sum_even_and_odd([])
+    {'positive_even': 0, 'positive_odd': 0, 'negative_even': 0, 'negative_odd': 0}
+    """
```
Comment (Azizsin7) [inline solutions/sum_even_and_odd.py:47]: add some defensive doctest
and 3 or more doctest

File: solutions/sum_even_and_odd.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            raise ValueError("All elements must be integers or floats.")
+
+    result = {
+        "positive_even": 0,
+        "positive_odd": 0,
+        "negative_even": 0,
+        "negative_odd": 0,
+    }
+
+    for num in numbers:
```
Comment (Azizsin7) [inline solutions/sum_even_and_odd.py:66]: your code does not handle the float numbers

**Review by Azizsin7 — CHANGES_REQUESTED** [review]
**For your function I did not see any comment** 

> 1. add
    - shebang line: #!/usr/bin/env python3
    - encoding declaration: # -- coding: utf-8 --
> 2. add annotations
> 3. add some defensive doctest and 3 or more doctest
> 4. your code does not handle the float numbers

**For your test file add some:**
> 1. Boundary testing with large lists
> 2. Defensive tests for invalid inputs like mixed types

File: solutions/sum_even_and_odd.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        "positive_odd": 0,
+        "negative_even": 0,
+        "negative_odd": 0,
+    }
+
+    # Iterate through the numbers and categorize them
+    for num in numbers:
+        if isinstance(num, (int, float)):
+            # Check if the number is positive and even
+            if num > 0 and float(num) % 2 == 0:
```
Comment (Azizsin7) [inline solutions/sum_even_and_odd.py]: change the float with int

if num > 0 and int(num) % 2 == 0:

File: solutions/sum_even_and_odd.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                result["positive_even"] += num
+            # Check if the number is positive and odd
+            elif num > 0:
+                result["positive_odd"] += num
+            # Check if the number is negative and even
+            elif num < 0 and float(num) % 2 == 0:
```
Comment (Azizsin7) [inline solutions/sum_even_and_odd.py]: change the float with int
elif num < 0 and int(num) % 2 == 0:

**Review by Azizsin7 — CHANGES_REQUESTED** [review]
change the float with int

**Review by Azizsin7 — APPROVED** [review]
Perfect
